File: THCS/2022_2023/PhuYen_2223/Bai3/BAI3.cpp
#include <iostream>
#include <fstream>
#include <stack>
#include <vector>

using namespace std;

vector<string> extract_expressions(const string& expression) {
    stack<int> positions;
    vector<string> results;
    
    for (int i = 0; i < expression.size(); i++) {
        if (expression[i] == '(') {
            // Khi gặp dấu mở ngoặc, đẩy vị trí vào ngăn xếp
            positions.push(i);
        }
        else if (expression[i] == ')') {
            // Khi gặp dấu đóng ngoặc, lấy vị trí dấu mở ngoặc gần nhất
            int start = positions.top();
            positions.pop();
            // Trích xuất biểu thức từ dấu mở ngoặc đến dấu đóng ngoặc
            results.push_back(expression.substr(start, i - start + 1));
        }
    }
    
    return results;
}

int main() {
    // Đọc dữ liệu từ file BAI3.INP
    ifstream inp_file("BAI3.INP");
    string expression;
    if (inp_file.is_open()) {
        getline(inp_file, expression);
        inp_file.close();
    } else {
        cerr << "Không thể mở file BAI3.INP" << endl;
        return 1;
    }

    // Tách các cụm biểu thức
    vector<string> expressions = extract_expressions(expression);

    // Ghi kết quả vào file BAI3.OUT
    ofstream out_file("BAI3.OUT");
    if (out_file.is_open()) {
        out_file << expressions.size() << endl;  // Ghi số lượng cụm
        for (const auto& expr : expressions) {
            out_file << expr << endl;  // Ghi từng cụm biểu thức
        }
        out_file.close();
    } else {
        cerr << "Không thể mở file BAI3.OUT" << endl;
        return 1;
    }

    return 0;
}

File: THCS/2022_2023/PhuYen_2223/Bai3/BAI3.py
def extract_expressions(expression):
    """
    Hàm tách các cụm biểu thức trong dấu ngoặc đơn.
    """
    stack = []
    results = []
    current_expression = []
    
    for i, char in enumerate(expression):
        if char == '(':
            # Nếu gặp dấu mở ngoặc, đẩy vị trí bắt đầu vào stack
            stack.append(i)
        elif char == ')':
            # Nếu gặp dấu đóng ngoặc, lấy vị trí mở ngoặc gần nhất
            start = stack.pop()
            # Trích xuất cụm biểu thức từ vị trí mở ngoặc đến vị trí đóng ngoặc
            current_expression = expression[start:i+1]
            results.append(current_expression)
    
    return results

# Đọc dữ liệu từ file BAI3.INP
with open("BAI3.INP", "r") as inp_file:
    expression = inp_file.readline().strip()

# Tách các cụm biểu thức
expressions = extract_expressions(expression)

# Ghi kết quả vào file BAI3.OUT
with open("BAI3.OUT", "w") as out_file:
    out_file.write(str(len(expressions)) + "\n")  # Ghi số lượng cụm
    for expr in expressions:
        out_file.write(expr + "\n")  # Ghi từng cụm

